README.md
librato-rack
=======

[![Gem Version](https://badge.fury.io/rb/librato-rack.png)](http://badge.fury.io/rb/librato-rack) [![Build Status](https://secure.travis-ci.org/librato/librato-rack.png?branch=master)](http://travis-ci.org/librato/librato-rack) [![Code Climate](https://codeclimate.com/github/librato/librato-rack.png)](https://codeclimate.com/github/librato/librato-rack)

---

## Important note on breaking change
**NOTE:** Starting with version 2.0.0 librato-rack requires a Librato account that [supports tagged metrics](https://www.librato.com/docs/kb/faq/account_questions/tags_or_sources/). 

If your Librato account doesn't yet support tagged metrics or you are using [a heroku addon](https://devcenter.heroku.com/articles/librato#using-with-ruby), please use the [1.x.x version](https://rubygems.org/gems/librato-rack/versions/1.1.0).

---

`librato-rack` provides rack middleware which will report key statistics for your rack applications to [Librato Metrics](https://metrics.librato.com/). It will also allow you to easily track your own custom metrics. Metrics are delivered asynchronously behind the scenes so they won't affect performance of your requests.

Currently Ruby 1.9.2+ is required.

## Upgrading

Upgrading from version 1.x to 2.x introduces breaking changes for legacy sources. Please contact [support@librato.com](mailto:support@librato.com) to migrate an existing Librato account.

## Quick Start

Install `librato-rack` as middleware in your application:

    use Librato::Rack

Configuring and relaunching your application will start the reporting of performance and request metrics. You can also track custom metrics by adding simple one-liners to your code:

    # keep counts of key events
    Librato.increment 'user.signup'

    # benchmark sections of code to verify production performance
    Librato.timing 'my.complicated.work' do
      # do work
    end

    # track averages across requests
    Librato.measure 'user.social_graph.nodes', user.social_graph.size

## Installation & Configuration

Install the gem:

    $ gem install librato-rack

Or add to your Gemfile if using bundler:

    gem "librato-rack"

In your rackup file or equivalent, require and add the middleware:

    require 'librato-rack'
    use Librato::Rack

In order to get the most accurate measurements, it is recommended that the `librato-rack` middleware be the first middleware in your stack. This will ensure that timing measurements like the `rack.request.time` metric will include all of the time spent in the Rack middleware stack.

If you don't have a Metrics account already, [sign up](https://metrics.librato.com/). In order to send measurements to Metrics you need to provide your account credentials to `librato-rack`. You can provide these one of two ways:

##### Use environment variables

By default you can use `LIBRATO_USER` and `LIBRATO_TOKEN` to pass your account data to the middleware. While these are the only required variables, there are a few more optional environment variables you may find useful.

* `LIBRATO_TAGS` - the default tags to use for submitted metrics. Format is comma-separated key=value pairs, e.g. `region=us-east,az=b`. If not set, `host` of the executing machine is detected and set as default tag
* `LIBRATO_SUITES` - manage which metrics librato-rack will report. See more in [metrics suites](#metric-suites).
* `LIBRATO_PREFIX` - a prefix which will be prepended to all metric names
* `LIBRATO_LOG_LEVEL` - see logging section for more
* `LIBRATO_PROXY` - HTTP proxy to use when connecting to the Librato API (can also use the `https_proxy` or `http_proxy` environment variable commonly supported by Linux command line utilities)
* `LIBRATO_AUTORUN` - set to `'0'` to prevent the reporter from starting, useful if you don't want `librato-rack` to start under certain circumstances
* `LIBRATO_EVENT_MODE` - use with evented apps, see "Use with EventMachine" below

##### Use a configuration object

If you want to do more complex configuration, use your own environment variables, or control your configuration in code, you can use a configuration object:

    config = Librato::Rack::Configuration.new
    config.user = 'myuser@mysite.com'
    config.token = 'mytoken'
    # …more configuration

    use Librato::Rack, :config => config

See the [configuration class](https://github.com/librato/librato-rack/blob/master/lib/librato/rack/configuration.rb) for all available options.

##### Running on Heroku

If you are using the [Librato Metrics Heroku addon](https://addons.heroku.com/librato), your `LIBRATO_USER` and `LIBRATO_TOKEN` environment variables will already be set in your Heroku environment. If you are running without the addon you will need to provide them yourself.

NOTE: if Heroku idles your application no measurements will be sent until it receives another request and is restarted. If you see intermittent gaps in your measurements during periods of low traffic this is the most likely cause.

## Default Tags

Librato Metrics supports tagged measurements that are associated with a metric, one or more tag pairs, and a point in time. For more information on tagged measurements, visit our [API documentation](https://www.librato.com/docs/api/#measurements).

##### Detected Tags

By default, `host` is detected and applied as a default tag for submitted measurements. Optionally, you can override the detected values, e.g. `LIBRATO_TAGS=host=myapp-prod-1`

##### Custom Tags

In addition to the default tags, you can also provide custom tags:

```ruby
config = Librato::Rack::Configuration.new
config.user = 'myuser@mysite.com'
config.token = 'mytoken'
config.tags = { service: 'myapp', environment: 'production', host: 'myapp-prod-1' }

use Librato::Rack, :config => config
```

##### Metric Suites

The metrics recorded by `librato-rack` are organized into named metric suites that can be selectively enabled/disabled:

* `rack`: The `rack.request.total`, `rack.request.time`, `rack.request.slow`, `rack.request.queue.time`, and `rack.processes` metrics
* `rack_status`: `rack.request.status` metric with `status` tag name and HTTP status code tag value, e.g. `status=200`
* `rack_method`: `rack.request.method` metric with `method` tag name and HTTP method tag value, e.g. `method=POST`

All three of the metric suites listed above are enabled by default.

The metric suites can be configured via either the `LIBRATO_SUITES` environment variable or the `suites` attributes on the `Librato::Rack::Configuration` object.

    LIBRATO_SUITES="rack,rack_method"  # use ONLY the rack and rack_method suites
    LIBRATO_SUITES="+foo,+bar"         # + prefix indicates that you want the default suites plus foo and bar
    LIBRATO_SUITES="-rack_status"      # - prefix indicates that you want the default suites removing rack_status
    LIBRATO_SUITES="+foo,-rack_status" # Use all default suites except for rack_status while also adding foo
    LIBRATO_SUITES="all"               # Enable all suites
    LIBRATO_SUITES="none"              # Disable all suites
    LIBRATO_SUITES=""                  # Use only the default suites (same as if env var is absent)

Note that you should EITHER specify an explict list of suites to enable OR add/subtract individual suites from the default list (using the +/- prefixes). If you try to mix these two forms a `Librato::Rack::InvalidSuiteConfiguration` error will be raised.

##### Use with EventMachine and EM Synchrony

`librato-rack` has experimental support for EventMachine and EM Synchrony apps.

When using in an evented context set LIBRATO_EVENT_MODE to `'eventmachine'` if using [EventMachine](https://github.com/eventmachine/eventmachine) or `'synchrony'` if using [EM Synchrony](https://github.com/igrigorik/em-synchrony) and/or [Rack::FiberPool](https://github.com/alebsack/rack-fiber_pool). We're interested in maturing this support, so please let us know if you have any issues.

## Custom Measurements

Tracking anything that interests you is easy with Metrics. There are four primary helpers available:

#### increment

Use for tracking a running total of something _across_ requests, examples:

```ruby
# increment the 'sales_completed' metric by one
Librato.increment 'sales.completed'
# => {:host=>"myapp-prod-1"}

# increment by five
Librato.increment 'items.purchased', by: 5
# => {:host=>"myapp-prod-1"}

# increment with custom per-measurement tags
Librato.increment 'user.purchases', tags: { user_id: user.id, currency: 'USD' }
# => {:user_id=>43, :currency=>"USD"}

# increment with custom per-measurement tags and inherited default tags
Librato.increment 'user.purchases', tags: { user_id: user.id, currency: 'USD' }, inherit_tags: true
# => {:host=>"myapp-prod-1", :user_id=>43, :currency=>"USD"}
```

Other things you might track this way: user signups, requests of a certain type or to a certain route, total jobs queued or processed, emails sent or received

###### Sporadic Increment Reporting

Note that `increment` is primarily used for tracking the rate of occurrence of some event. Given this increment metrics are _continuous by default_: after being called on a metric once they will report on every interval, reporting zeros for any interval when increment was not called on the metric.

Especially with custom per-measurement tags you may want the opposite behavior - reporting a measurement only during intervals where `increment` was called on the metric:

```ruby
# report a value for 'user.uploaded_file' only during non-zero intervals
Librato.increment 'user.uploaded_file', tags: { user_id: user.id, bucket: bucket.name }, sporadic: true
```

#### measure

Use when you want to track an average value _per_-request. Examples:

    Librato.measure 'user.social_graph.nodes', 212

    # report from custom per-measurement tags
    Librato.measure 'jobs.queued', 3, tags: { priority: 'high', worker: 'worker.12' }

#### timing

Like `Librato.measure` this is per-request, but specialized for timing information:

    Librato.timing 'twitter.lookup.time', 21.2

The block form auto-submits the time it took for its contents to execute as the measurement value:

    Librato.timing 'twitter.lookup.time' do
      @twitter = Twitter.lookup(user)
    end

#### percentiles

By defaults timings will send the average, sum, max and min for every minute. If you want to send percentiles as well you can specify them inline while instrumenting:

```ruby
# track a single percentile
Librato.timing 'api.request.time', time, percentile: 95

# track multiple percentiles
Librato.timing 'api.request.time', time, percentile: [95, 99]
```

You can also use percentiles with the block form of timings:

```ruby
Librato.timing 'my.important.event', percentile: 95 do
  # do work
end
```

#### group

There is also a grouping helper, to make managing nested metrics easier. So this:

    Librato.measure 'memcached.gets', 20
    Librato.measure 'memcached.sets', 2
    Librato.measure 'memcached.hits', 18

Can also be written as:

    Librato.group 'memcached' do |g|
      g.measure 'gets', 20
      g.measure 'sets', 2
      g.measure 'hits', 18
    end

Symbols can be used interchangeably with strings for metric names.

## Use with Background Workers / Cron Jobs

`librato-rack` is designed to run within a long-running process and report periodically. Intermittently running rake tasks and most background job tools (delayed job, resque, queue_classic) don't run long enough for this to work.

Never fear, [we have some guidelines](https://github.com/librato/librato-rails/wiki/Monitoring-Background-Workers) for how to instrument your workers properly.

If you are using `librato-rack` with sidekiq, [see these notes about setup](https://github.com/librato/librato-rails/wiki/Monitoring-Background-Workers#monitoring-long-running-threaded-workers-sidekiq-etc).

## Cross-Process Aggregation

`librato-rack` submits measurements back to the Librato platform on a _per-process_ basis. By default these measurements are then combined into a single measurement per default tags (detects `host`) before persisting the data.

For example if you have 4 hosts with 8 unicorn instances each (i.e. 32 processes total), on the Metrics site you'll find 4 data streams (1 per host) instead of 32.
Current pricing applies after aggregation, so in this case you will be charged for 4 streams instead of 32.

## Troubleshooting

Note that it may take 2-3 minutes for the first results to show up in your Metrics account after you have started your servers with `librato-rack` enabled and the first request has been received.

For more information about startup and submissions to the Metrics service you can set your `log_level` to `debug`. If you are having an issue with a specific metric, using `trace` will add the exact measurements being sent to your logs along with other details about `librato-rack` execution. Neither of these modes are recommended long-term in production as they will add significant volume to your log file and may slow operation somewhat.

Submission times are total time but submission I/O is non-blocking - your process will continue to handle requests during submissions.

If you are debugging setup locally you can set `flush_interval` to something shorter (say 10s) to force submission more frequently. Don't change your `flush_interval` in production as it will not result in measurements showing up more quickly, but may affect performance.

## Contribution

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project and submit a pull request from a feature or bugfix branch.
* Please include tests. This is important so we don't break your changes unintentionally in a future version.
* Please don't modify the gemspec, Rakefile, version, or changelog. If you do change these files, please isolate a separate commit so we can cherry-pick around it.

## Copyright

Copyright (c) 2013-2017 [Librato Inc.](http://librato.com) See LICENSE for details.


CHANGELOG.md
### Version 2.0.3
* Fix bug to prevent `rack.processes` reporting when `rack` suite is disabled (#75, Sam Moore)

### Version 2.0.2
* Update `ValidatingQueue` tag value restrictions (#72)

### Version 2.0.1
* Ensure `rack.request.queue.time` metric is not negative (#68, Austin Schneider)

### Version 2.0.0
* Add support for tagged measurements (#54). **NOTE**: This version introduces breaking changes for legacy sources. Please contact support@librato.com to learn more.

### Version 1.1.0
* Fix deprecation warnings in ruby 2.4 (#57, Ben Radler)

### Version 1.0.1
* Fix missing p95 for rack.request.time

### Version 1.0.0
* Add support for configurable metric suites
* Drop support for long-deprecated config via LIBRATO_METRICS_* env vars
* Drop support for old-style config passing during initialization
* Deprecate `disable_rack_metrics` config option, use `suites='none'` instead

### Version 0.6.0
* Add support for proxy configuration

### Version 0.5.0
* Add support for percentiles when timing
* Report p95 for rack.request.time and rack.request.queue.time

### Version 0.4.5
* Add #start! to tracker

### Version 0.4.4
* Relax version constraint for librato-metrics

### Version 0.4.3
* Update queue wait support to tolerate float-style timestamps

### Version 0.4.2
* Move gem sign code to rake task, fixes bug bundling in some environments

### Version 0.4.1
* Support a pre-configured tracker object
* Make log-prefix configurable
* Break pid-locking out of startup checks

### Version 0.4.0
* Add HTTP method (GET, POST) metrics
* Add log buffering support
* Ensure all options passed to a grouped increment are respected
* LIBRATO_AUTORUN can be used to prevent startup
* Add ability to interrupt reporter process
* Start reporting deprecations for old config methods
* Add docs for best practices for background workers
* Other documentation improvements

### Version 0.3.0
* Add experimental support for EventMachine and EMSynchrony (Balwant K)
* Start testing suite against jruby/rbx
* Gem is now signed

### Version 0.2.1
* Fix exception if logging metrics before middleware init (Eric Holmes)

### Version 0.2.0
* Add disable_rack_metrics config option
* Remove metrics based on deprecated heroku HTTP headers
* Ensure compatibility with ruby 2.0

### Version 0.1.0
* Initial version


lib/librato/rack.rb
require 'thread'
require 'librato/metrics'

module Librato
  extend SingleForwardable
  def_delegators :tracker, :increment, :measure, :timing, :group

  def self.register_tracker(tracker)
    @tracker = tracker
  end

  def self.tracker
    @tracker ||= Librato::Rack::Tracker.new(Librato::Rack::Configuration.new)
  end
end

module Librato
  # Middleware for rack applications. Installs tracking hearbeat for
  # metric submission and tracks performance metrics.
  #
  # @example A basic rack app
  #   require 'rack'
  #   require 'librato-rack'
  #
  #   app = Rack::Builder.app do
  #     use Librato::Rack
  #     run lambda { |env| [200, {"Content-Type" => 'text/html'}, ["Hello!"]] }
  #   end
  #
  # @example Using a custom config object
  #   config = Librato::Rack::Configuration.new
  #   config.user = 'myuser@mysite.com'
  #   config.token = 'mytoken'
  #   …more configuration
  #
  #   use Librato::Rack, :config => config
  #   run MyApp
  #
  class Rack
    RECORD_RACK_BODY = <<-'EOS'
      group.increment 'total'
      group.timing    'time', duration, percentile: 95
      group.increment 'slow' if duration > 200.0
    EOS

    RECORD_RACK_STATUS_BODY = <<-'EOS'
      status_tags = { status: status }
      tracker.increment "rack.request.status", tags: status_tags, inherit_tags: true
      tracker.timing "rack.request.status.time", duration, tags: status_tags, inherit_tags: true
    EOS

    RECORD_RACK_METHOD_BODY = <<-'EOS'
      method_tags = { method: http_method.downcase! }
      tracker.increment "rack.request.method", tags: method_tags, inherit_tags: true
      tracker.timing "rack.request.method.time", duration, tags: method_tags, inherit_tags: true
    EOS

    attr_reader :config, :tracker

    def initialize(app, options={})
      @app = app
      @config = options.fetch(:config, Configuration.new)
      @tracker = @config.tracker || Tracker.new(@config)
      Librato.register_tracker(@tracker) # create global reference

      build_record_request_metrics_method
      build_record_header_metrics_method
      build_record_exception_method
    end

    def call(env)
      check_log_output(env) unless @log_target
      @tracker.check_worker
      record_header_metrics(env)
      response, duration = process_request(env)
      record_request_metrics(response.first, env["REQUEST_METHOD"], duration)
      response
    end

    private

    # this generally will only get called on the first request
    # it figures out the environment-appropriate logging outlet
    # and notifies config and tracker about it
    def check_log_output(env)
      return if @log_target
      if in_heroku_env?
        tracker.on_heroku = true
        default = ::Logger.new($stdout)
      else
        default = env['rack.errors'] || $stderr
      end
      @tracker.update_log_target(config.log_target ||= default)
      @log_target = config.log_target
    end

    def in_heroku_env?
      # don't have any custom http vars anymore, check if hostname is UUID
      Socket.gethostname =~ /[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}/i
    end

    def process_request(env)
      time = Time.now
      begin
        response = @app.call(env)
      rescue Exception => e
        record_exception(e)
        raise
      end
      duration = (Time.now - time) * 1000.0
      [response, duration]
    end


    # Dynamically construct :record_request_metrics method based on
    # configured metric suites
    def build_record_request_metrics_method
      body = "def record_request_metrics(status, http_method, duration)\n"
      body << "return if config.disable_rack_metrics\n"

      unless config.instance_of?(Librato::Rack::Configuration::SuitesNone)
        body << "tracker.group 'rack.request' do |group|\n"

        if tracker.suite_enabled?(:rack)
          body << RECORD_RACK_BODY
        end

        if tracker.suite_enabled?(:rack_status)
          body << RECORD_RACK_STATUS_BODY
        end

        if tracker.suite_enabled?(:rack_method)
          body << RECORD_RACK_METHOD_BODY
        end

        body << "end\n"
      end

      body << "end\n"

      instance_eval(body)
    end

    # Dynamically construct :record_header_metrics method based on
    # configured metric suites
    def build_record_header_metrics_method
      if tracker.suite_enabled?(:rack)
        define_singleton_method(:record_header_metrics) do |env|
          queue_start = env['HTTP_X_REQUEST_START'] || env['HTTP_X_QUEUE_START']
          if queue_start
            queue_start = queue_start.to_s.sub('t=', '').sub('.', '')
            case queue_start.length
            when 16 # microseconds
              wait = ((Time.now.to_f * 1000000).to_i - queue_start.to_i) / 1000.0
              wait = 0 if wait < 0 # make up for potential time drift between the routing server and the application server
              tracker.timing 'rack.request.queue.time', wait, percentile: 95
            when 13 # milliseconds
              wait = (Time.now.to_f * 1000).to_i - queue_start.to_i
              wait = 0 if wait < 0 # make up for potential time drift between the routing server and the application server
              tracker.timing 'rack.request.queue.time', wait, percentile: 95
            end
          end
        end
      else
        define_singleton_method(:record_header_metrics) do |env|
          # no-op
        end
      end
    end

    def build_record_exception_method
      if tracker.suite_enabled?(:rack)
        define_singleton_method(:record_exception) do |exception|
          return if config.disable_rack_metrics
          tracker.increment 'rack.request.exceptions'
        end
      else
        define_singleton_method(:record_exception) do |exception|
          # no-op
        end
      end
    end

  end
end

require 'librato/collector'
require 'librato/rack/configuration'
require 'librato/rack/errors'
require 'librato/rack/logger'
require 'librato/rack/tracker'
require 'librato/rack/validating_queue'
require 'librato/rack/version'
require 'librato/rack/worker'
